surmap/3d_shape.cpp

```diff
@@ -0,0 +1,757 @@
+		}
+
+	if(XKey.Pressed(SDLK_UP))
+		if(alt)
+			shape_scale_y *= scale_step;
+		else
+			shape.y_off -= move_step;
+	if(XKey.Pressed(SDLK_DOWN))
+		if(alt)
+			shape_scale_y /= scale_step;
+		else
+			shape.y_off += move_step;
+	if(XKey.Pressed(SDLK_LEFT))
+		if(alt)
+			shape_scale_x /= scale_step;
+		else
+			shape.x_off -= move_step;
+	if(XKey.Pressed(SDLK_RIGHT))
+		if(alt)
+			shape_scale_x *= scale_step;
+		else
+			shape.x_off += move_step;
+
+#ifdef _SURMAP_
+	shape.x_off &= clip_mask_x;
+	shape.y_off &= clip_mask_y;
+#endif
+
+	return 1;
+}
+
+void load_shape(char* name,int x,int y)
+{
+	XStream c3d(name,XS_IN);
+	XBuffer buf(c3d.size());
+	c3d.read(buf.address(),c3d.size());
+	c3d.close();
+	shape.loadC3D(buf);
+
+	shape.x_off = x;
+	shape.y_off = y;
+	shape.z_off = 128;
+
+	catch_log = 1;
+
+#ifdef _SHAPE_
+	upper_buffer = new unsigned char[1024*1024];
+	lower_buffer = new unsigned char[1024*1024];
+#endif
+}
+#ifndef _SURMAP_
+void prepare_shape(int mode,char* name,int new_size)
+{
+	char* c3d_name = ALLOC(strlen(name) + 1,char);
+	strlwr(name);
+	strcpy(c3d_name,name);
+	strcpy(strstr(c3d_name,"asc"),"c3d");
+
+	if(mode & SHAPE_QUIT_IF_EXIST){
+		XStream c3d(0);
+		if(c3d.open(c3d_name,XS_IN)){
+			int version;
+			c3d > version;
+			if(version == C3D_VERSION_3){
+				XCon < "\nFile \"" < name < "\" has already been converted.\n";
+				return;
+				}
+			}
+		}
+		
+	shape.loadASC(name);
+
+	double scale_mul = 1;
+	if(mode & SHAPE_SCALE)
+		scale_mul = new_size/(double)(mode & SHAPE_RELATIVE ? 100 : shape.rmax);
+
+	Vector offset = Vector(shape.x_off,shape.y_off,shape.z_off);
+	if(mode & SHAPE_LOAD_PARAMS){
+		XStream f(0);
+		if(f.open("shape.prm",XS_IN)){
+			Parser prm_in("shape.prm");
+			prm_in.search_name("scale_mul:");
+			scale_mul = prm_in.get_int();
+			
+			prm_in.search_name("x_offset:");
+			offset.x = prm_in.get_int();
+			prm_in.search_name("y_offset:");
+			offset.y = prm_in.get_int();
+			offset.z = 0;
+			}
+		}
+
+	XStream prm_out("shape.prm",XS_OUT);
+	prm_out.SetDigits(8);
+
+	XCon < "\nConverting ASC to C3D: " < name < "\n";
+	prm_out < "\nConverting ASC to C3D: " < name < "\n";
+	
+	XCon < "Faces: " <= shape.num_poly < "\n";
+	prm_out < "Faces: " <= shape.num_poly < "\n";
+	XCon < "Vertices: " <= shape.num_vert < "\n";
+	prm_out < "Vertices: " <= shape.num_vert < "\n";
+
+	XCon < "Original bound radius: " <= shape.rmax < "\n";
+	prm_out < "Original bound radius: " <= shape.rmax < "\n";
+
+
+	shape.set_colors(COLORS_IDS::BODY,0,0);
+
+	if(mode & SHAPE_MOVE)
+		shape.move(-offset);
+	shape.make_normals();
+	shape.pre_sorting();
+	shape.scale(scale_mul,1);
+
+	XCon < "Obtained bound radius: " <= shape.rmax < "\n";
+	prm_out < "Obtained bound radius: " <= shape.rmax < "\n";
+
+	prm_out < "scale_mul: " <= scale_mul < "\n";
+	prm_out < "x_offset: " <= offset.x < "\n";
+	prm_out < "y_offset: " <= offset.y < "\n";
+
+	XStream fout(c3d_name,XS_OUT);
+	shape.saveC3D(fout);
+
+	delete c3d_name;

```

Comment:
delete[] ?

Reply:
I believe should be fixed in master branch, because this code is disabled by #ifndef